tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+	hrw := &HeadReadWriter{
+		dirFile:     dirFile,
+		n:           0,
+		segmentTime: segmentTime,
+		pool:        pool,
+	}
+
+	if err := hrw.initReader(); err != nil {
+		return nil, err
+	}
+
+	return hrw, nil
+}
+
+// NewDirReader returns a new Reader against sequentially numbered files in the
+// given directory.
+func (w *HeadReadWriter) initReader() (err error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
+	defer func() {
+		if err != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(err)
+			merr.Add(closeAllFromMap(cs))
+			err = merr
+		}
+	}()
+
+	files, err := sequenceFilesMap(w.dirFile.Name())
+	if err != nil {
+		return err
+	}
+	if w.pool == nil {
+		w.pool = chunkenc.NewPool()
+	}
+
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
+	}
+	w.bs = bs
+	w.cs = cs
+
+	w.size = 0
+	for i, b := range w.bs {
+		if b.Len() < HeadSegmentHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid segment header in segment %d", i)
+		}
+		// Verify magic number.
+		if m := binary.BigEndian.Uint32(b.Range(0, MagicChunksSize)); m != MagicChunks {
+			return errors.Errorf("invalid magic number %x", m)
+		}
+
+		// Verify chunk format version.
+		if v := int(b.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+8))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last segment. We set it to current time so that it will be eventually deleted.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			binary.BigEndian.PutUint64(w.buf[:], uint64(time.Now().Unix()))
+			if _, err := f.WriteAt(w.buf[:8], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		w.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func sequenceFilesMap(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+func closeAllFromMap(cs map[int]io.Closer) error {
+	var merr tsdb_errors.MultiError
+
+	for _, c := range cs {
+		merr.Add(c.Close())
+	}
+	return merr.Err()
+}
+
+// IterateAllChunks iterates on all the chunks in it's byte slices in the order of the segment file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (w *HeadReadWriter) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	// Iterate files in ascending order.
```

Comment:
Do we check for gaps?

Reply:
We do not. The gap could be for many reasons and I think we should not check for gaps.

Reply:
What sort of reasons? Gaps aren't possible for the normal WAL, and we check for gaps. I don't see why we wouldn't be doing the same reason here.

Reply:
I think I misunderstood what a gap is here, so, what do you mean by gap here?

Reply:
If the files are say 1, 3 and 4. 2 is missing.

Reply:
Ahh, yes. I definitely misunderstood before. I will make it error out in that case. Once https://github.com/prometheus/prometheus/pull/6129 is in, I can follow up with a similar logic.